feat(influx_tools): Add export to parquet files

srebhan opened the pull request:
Closes #

Describe your proposed changes here.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

This PR adds a command to export data into per-shard parquet files. To do so, the command iterates over the shards, creates a cumulative schema over the series of a measurement (i.e. a super-set of tags and fields) and exports the data to a parquet file per measurement and shard.

To test the tool run
``` 
go run -ldflags "-X google.golang.org/protobuf/reflect/protoregistry.conflictPolicy=ignore" ./cmd/influx_tools/ export-parquet -config influxdb.conf -database telegraf
```

davidby-influx commented on cmd/influx_tools/parquet/exporter.go:
```diff
@@ -0,0 +1,488 @@
+	return e.store.Close()
+}
+
+func (e *exporter) printPlan() {
```
This can fail and maybe should return an error

srebhan replied:
How can this fail? If you are taking about the error when calling `creator.validate()`, then this is expected as this will just signal that the schema is not valid and has unresolved conflicts. The actual error handling is then done in `export()`...

davidby-influx commented on cmd/influx_tools/parquet/exporter.go:
```diff
@@ -0,0 +1,488 @@
+				return fmt.Errorf("exporting measurement %q in shard %d failed: %w", measurement, shard.ID(), err)
+			}
+		}
+		fmt.Fprintf(e.writer, "Finished export of shard %d in %v...\n", shard.ID(), time.Since(start))
```
maybe use a logger here?

davidby-influx commented on cmd/influx_tools/parquet/exporter.go:
```diff
@@ -0,0 +1,488 @@
+	if err != nil {
+		return fmt.Errorf("creating file %q failed: %w", filename, err)
+	}
+	defer file.Close()
```
`Capture` error return from `Close`

davidby-influx commented on cmd/influx_tools/parquet/exporter.go:
```diff
@@ -0,0 +1,488 @@
+	if err != nil {
+		return fmt.Errorf("creating parquet writer for file %q failed: %w", filename, err)
+	}
+	defer writer.Close()
```
`Capture` error from `Close`.  A frequent source fo these errors is the disk being full.

davidby-influx replied:
`Capture` error from `Close`.  A frequent source fo these errors is the disk being full.

davidby-influx commented on cmd/influx_tools/parquet/schema.go:
```diff
@@ -0,0 +1,273 @@
+		if err != nil {
+			return fmt.Errorf("getting series cursor failed: %w", err)
+		}
+		defer seriesCursor.Close()
```
`Capture` error

davidby-influx commented on cmd/influx_tools/parquet/schema.go:
```diff
@@ -0,0 +1,273 @@
+			s.fieldKeys = append(s.fieldKeys, name)
+		default:
+			// Unconvertible type
+			panic(fmt.Errorf("unconvertible field %q with type %v", name, ftype))
```
panic or better to return an error?  Or ignore that field and try to continue?

srebhan replied:
Decided to return an error. I don't want to start out too permissive so let's error out here for now and decide for a more graceful error handling if need be.

davidby-influx reviewed (commented):
I did a quick review, but I'm not familiar with arrow and certainly missed some things.  I can do a more thorough review if we paired to walk through the algorithm once.

srebhan commented on .circleci/config.yml:
```diff
@@ -6,7 +6,7 @@ parameters:
     description: docker tag for cross build container from quay.io . Created by https://github.com/influxdata/edge/tree/master/dockerfiles/cross-builder .
     type: string
-    default: go1.21.12-latest
+    default: go1.21.10-latest
```
Unintentional. Reverted.

srebhan commented on cmd/influx_tools/parquet/batcher.go:
```diff
@@ -0,0 +1,236 @@
+	converter map[string]func(interface{}) (interface{}, error),
+	nameMapping map[string]string,
+) (*batcher, error) {
+	seriesCursor, err := shard.CreateSeriesCursor(
```
Left-over from a previous version. Removed.

srebhan commented on cmd/influx_tools/parquet/batcher.go:
```diff
@@ -0,0 +1,236 @@
+				cursor.Close()
+				panic(fmt.Errorf("unexpected type %T", cursor))
+			}
+			cursor.Close()
```
No it does not.

srebhan commented on cmd/influx_tools/parquet/batcher.go:
```diff
@@ -0,0 +1,236 @@
+			}
+			fieldEnd := models.MaxNanoTime
+			switch c := cursor.(type) {
+			case tsdb.IntegerArrayCursor:
```
I tried my very best, but due to the fact that each implementation returns a different type and function signature for `Next()` I could not find a way to make this generic. Do you have some idea?

davidby-influx replied:
I figured out how to have implement a generic `Next` function, but the access to the Values and Timestamps fields are the problem. Even with type constraints that all have two fields named `Values` and `Timestamps` the compiler inference fails.
https://goplay.tools/snippet/_BsjhhcYGGB

srebhan commented on cmd/influx_tools/parquet/batcher.go:
```diff
@@ -0,0 +1,236 @@
+				for i, t := range values.Timestamps {
+					v, err := converter(values.Values[i])
+					if err != nil {
+						fmt.Fprintf(os.Stderr, "converting %v of field %q failed: %v", values.Values[i], field, err)
```
Done.

srebhan commented:
@davidby-influx thanks for the thorough review! I tried to address all issues and commented on the three unresolved ones. Will schedule a meeting for walking through the code. Thanks again!

davidby-influx commented on cmd/influx_tools/parquet/batcher.go:
```diff
@@ -0,0 +1,236 @@
+	}
+
+	// Extract the rows ordered by timestamp
+	var rows []row
```
make([]row, ....) some useful value.

davidby-influx commented on cmd/influx_tools/parquet/cursors.go:
This is cleaner; I can't think of a way to make the valueCursor implementation generic, so this seems like the best we can do.

davidby-influx reviewed (approved):
LGTM

alespour commented:
I'm not sure what to make of this: I have v1 db with several measurements like cpu disk etc, each with ~8M rows

```
> select count(usage_user) from cpu

name: cpu
time count
---- -----
0    8631360
```

The same query returns different number of rows in exported Parquet "db":

```
alespour@master-node:/bigdata/x$ duckdb -column -s "select count(usage_user) from 'all/cpu-*.parquet'"

count(usage_user)
-----------------
28771200 
```

alespour commented:
* tested measurement without tags - OK
* tested single & all measurements export - appears OK, _except the discrepancy of number of rows_

Tested with db with simulating 1-month of monitoring data of a small data center (9 measurements like cpu, disk etc, 10 tags). DB files size on disk 4.1 GB, 5 shards.

Exported Parquet size on disk 11 GB, took 1h6m on somewhat obsolete laptop (Core i7 CPU, 8-core, 16 GB RAM, SSD). Memory usage during export was stable (RSS peak ~2 GB).

InfluxDB measuement structure example:
```
> show tag keys from cpu
name: cpu
tagKey
------
arch
datacenter
hostname
os
rack
region
service
service_environment
service_version
team

> show field keys from cpu
name: cpu
fieldKey         fieldType
--------         ---------
usage_guest      float
usage_guest_nice float
usage_idle       float
usage_iowait     float
usage_irq        float
usage_nice       float
usage_softirq    float
usage_steal      float
usage_system     float
usage_user       float
```

Parquet:
```
alespour@master-node:/bigdata/x$ duckdb -column -s "describe select * from 'all/cpu-*.parquet'"

column_name          column_type  null  key  default  extra
-------------------  -----------  ----  ---  -------  -----
time                 TIMESTAMP    YES                      
arch                 VARCHAR      YES                      
datacenter           VARCHAR      YES                      
hostname             VARCHAR      YES                      
os                   VARCHAR      YES                      
rack                 VARCHAR      YES                      
region               VARCHAR      YES                      
service              VARCHAR      YES                      
service_environment  VARCHAR      YES                      
service_version      VARCHAR      YES                      
team                 VARCHAR      YES                      
usage_guest          DOUBLE       YES                      
usage_guest_nice     DOUBLE       YES                      
usage_idle           DOUBLE       YES                      
usage_iowait         DOUBLE       YES                      
usage_irq            DOUBLE       YES                      
usage_nice           DOUBLE       YES                      
usage_softirq        DOUBLE       YES                      
usage_steal          DOUBLE       YES                      
usage_system         DOUBLE       YES                      
usage_user           DOUBLE       YES                      

alespour@master-node:/bigdata/x$ duckdb -column -s "select count(*) from 'all/cpu-*.parquet'"

count_star()
------------
28771200    
```

alespour commented:
My apologies, it was a mistake on my side. Row count matches.

InfluxDB:
```
> select count(usage_user) from cpu
name: cpu
time count
---- -----
0    28771200
```

Parquet:
```
alespour@master-node:/bigdata/x$ duckdb -column -s "select count(usage_user) from 'cpu/*.parquet'"
count(usage_user)
-----------------
28771200  
```

alespour commented:
It's GTG by me :+1:

srebhan commented:
To run the exporter in this PR do the following
1. Clone the repo and checkout the PR
```shell
# git clone https://github.com/influxdata/influxdb.git
# cd influxdb/
# git fetch origin pull/25297/head:v1-bulk-exporter-parquet 
# git checkout v1-bulk-exporter-parquet
```
2. Build InfluxDB v1
```shell
# export PKG_CONFIG=${PWD}/pkg-config.sh
# go build ./...
```
3. Run the exporter (with the help flag)
```shell
# go run -ldflags "-X google.golang.org/protobuf/reflect/protoregistry.conflictPolicy=ignore" ./cmd/influx_tools/ export-parquet --help
```
4. Run the exporter with the config of an existing server instance
```shell
# go run -ldflags "-X google.golang.org/protobuf/reflect/protoregistry.conflictPolicy=ignore" ./cmd/influx_tools/ export-parquet -config <path to influxdb config dir>/influxdb.conf -database <database to export>
```

dburton-influxdata commented:
Do we have a compiled version to test with or do I still need to clone the repo and build the go binary?

dburton-influxdata commented:
Here is the Python script in Zip format for Github.
[exporter_script.zip](https://github.com/user-attachments/files/17806909/exporter_script.zip)

srebhan commented:
@dburton-influxdata using `os.environ` does __NOT__ export the variable to subprocesses like the go command! You would need to use [`os.putenv`](https://docs.python.org/3/library/os.html#os.putenv) but I don't understand why you need to use python for the whole thing...

srebhan commented:
Thanks for your investigations @jwei-influx! Let me answer your questions:

> 1. Has there been any consideration about how the tool is intended to handle mixed field type shards?

Yes. There are the `--resolve-types` and `--resolve-names` command-line options to fix type conflicts and name conflicts (between tags and fields) respectively.

> 2. If we need to do any sort of custom partitioning on the eventual 3.0 system, are we able to do that with this tool? Or conversely, is the resulting parquet file from this tool able to be slotted in behind a custom partitioning scheme that is pre-applied to the 3.0 instance?

This is beyond the scope of this tool. This tool outputs the data as-is (plus potential type and name changes as discussed above) without the ability to split, merge or modify data.

> 3. Are we able to do any sort of manipulation of the tags and fields using this tool, or potentially by editing the resulting parquet files?

As I mentioned above, you can modify field names (and types) using the `--resolve-types` and `--resolve-names` command-line options. Beyond this, the tool is not intended for manipulating or editing data or schemata but for _exporting_ of existing data!

> 4. Are we able to use this for backloading processes? ie: slotting the resulting parquet files into an existing database that's receiving the real-time dual-written feed from the original 1.x system

This is a question you should ask the authors of the import tool that takes the parquet files generated by this tool! Maybe @jacobmarble can answer this question or knows who could answer this...

> I might have more questions as I test the tool, but these are the ones that are top of mind for me right now.

Happy to answer them. ;-)

jacobmarble commented:
> > 4. Are we able to use this for backloading processes? ie: slotting the resulting parquet files into an existing database that's receiving the real-time dual-written feed from the original 1.x system
> 
> This is a question you should ask the authors of the import tool that takes the parquet files generated by this tool! Maybe @jacobmarble can answer this question or knows who could answer this...

I haven't looked closely at this PR, and I'm no longer managing the team that is working on Parquet import. @helenosheaa might direct the right person to answer thoughtfully and accurately.

barbaranelson commented:
Adding a reference to the bulk import doc, in case it's useful:
https://docs.google.com/document/d/1-KUAKLAWfzHmEVRPKKF-jo7YL2gLEFT1MXa5tqyFm-4/edit?usp=sharing

helenosheaa commented:
I missed the @ from jacob here. 

So this is tooling so that you can export from v1 to parquet and the question is whether that will line up with the bulk_ingester which takes parquet files - there's docs of how the bulk_ingester currently functions [here](https://github.com/influxdata/influxdb_iox/tree/main/bulk_ingester) and its restrictions 

@carols10cents or @stuartcarnie might be good additional reviewers if needed

carols10cents commented:
> > > 7\. Are we able to use this for backloading processes? ie: slotting the resulting parquet files into an existing database that's receiving the real-time dual-written feed from the original 1.x system
> > 
> > 
> > This is a question you should ask the authors of the import tool that takes the parquet files generated by this tool! Maybe @jacobmarble can answer this question or knows who could answer this...
> 
> I haven't looked closely at this PR, and I'm no longer managing the team that is working on Parquet import. @helenosheaa might direct the right person to answer thoughtfully and accurately.

The `bulk_ingester` tool is able to import data into tables that are also being written to directly; the compactor will reorganize and deduplicate as needed. The `bulk_ingester` also repartitions data according to the partition template set on the destination table. In other words, it rewrites the Parquet files during the import, which does take some time-- the Parquet files are not added to the system directly as they are.

Please let me know if you have any other questions!

srebhan commented:
@jwei-influx I rebased the PR and fixed the help. You can now checkout this PR and then just run

```shell
# go build -ldflags "-X google.golang.org/protobuf/reflect/protoregistry.conflictPolicy=ignore" ./cmd/influx_tools/  
```

to build the `influx_tools` binary. Afterwards you can check the help using

```shell
# ./influx_tools export-parquet --help
```
and experiment with the tool like

```shell
# ./influx_tools export-parquet -config <path to influxdb config dir>/influxdb.conf -database <database to export>
```

Let me know if there are any issues!

davidby-influx reviewed (approved):
LGTM

jcdauchy-moodys commented:
@srebhan  Do you think it would be possible to filter some data by shard or even date as it is quite slow to export a large database ?

Thanks it works quite well.
